File: ios/Memorezar/Data/Services/AuthService.swift
import Foundation
import AuthenticationServices
import Combine

/// Handles authentication with Supabase via Apple, Google, and email/password
class AuthService: NSObject, ObservableObject {
    static let shared = AuthService()

    @Published var currentUser: AuthUser?
    @Published var isLoading = false
    @Published var errorMessage: String?

    var isSignedIn: Bool { currentUser != nil }

    var accessToken: String? {
        KeychainHelper.loadString(key: "access_token")
    }

    private var refreshToken: String? {
        KeychainHelper.loadString(key: "refresh_token")
    }

    private var refreshTimer: Timer?
    private let session = URLSession.shared

    private override init() {
        super.init()
    }

    // MARK: - Session Restore

    func restoreSession() {
        guard let accessToken = accessToken else { return }

        // Decode user from stored data
        if let userData = KeychainHelper.load(key: "user_data"),
           let user = try? JSONDecoder().decode(AuthUser.self, from: userData) {
            currentUser = user
            startRefreshTimer()

            // Silently refresh if token might be stale
            Task {
                await refreshTokenIfNeeded()
            }
        } else {
            // Have token but no user data — try to refresh
            Task {
                await refreshTokenIfNeeded()
            }
        }
    }

    // MARK: - Email/Password

    func signUpWithEmail(email: String, password: String, name: String) async throws {
        let url = URL(string: "\(SupabaseConfig.projectURL)/auth/v1/signup")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue(SupabaseConfig.anonKey, forHTTPHeaderField: "apikey")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: Any] = [
            "email": email,
            "password": password,
            "data": ["display_name": name]
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw AuthError.networkError
        }

        if httpResponse.statusCode == 200 || httpResponse.statusCode == 201 {
            // Some Supabase configs require email confirmation
            let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
            if let accessToken = json?["access_token"] as? String {
                try handleAuthResponse(json: json!, displayName: name)
            } else {
                // Email confirmation required
                throw AuthError.confirmationRequired
            }
        } else {
            let errorJson = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
            let msg = errorJson?["msg"] as? String
                ?? errorJson?["error_description"] as? String
                ?? "Sign up failed"
            throw AuthError.serverError(msg)
        }
    }

    func signInWithEmail(email: String, password: String) async throws {
        let url = URL(string: "\(SupabaseConfig.projectURL)/auth/v1/token?grant_type=password")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue(SupabaseConfig.anonKey, forHTTPHeaderField: "apikey")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body: [String: String] = ["email": email, "password": password]
        request.httpBody = try JSONEncoder().encode(body)

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw AuthError.networkError
        }

        if (200...299).contains(httpResponse.statusCode) {
            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                throw AuthError.invalidResponse
            }
            try handleAuthResponse(json: json)
        } else {
            let errorJson = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
            let msg = errorJson?["msg"] as? String
                ?? errorJson?["error_description"] as? String
                ?? "Invalid email or password"
            throw AuthError.serverError(msg)
        }
    }

    // MARK: - Apple Sign-In

    func signInWithApple(idToken: String, nonce: String?) async throws {
        let url = URL(string: "\(SupabaseConfig.projectURL)/auth/v1/token?grant_type=id_token")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue(SupabaseConfig.anonKey, forHTTPHeaderField: "apikey")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        var body: [String: String] = [
            "provider": "apple",
            "id_token": idToken
        ]
        if let nonce = nonce {
            body["nonce"] = nonce
        }
        request.httpBody = try JSONEncoder().encode(body)

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw AuthError.networkError
        }

        if (200...299).contains(httpResponse.statusCode) {
            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                throw AuthError.invalidResponse
            }
            try handleAuthResponse(json: json)
        } else {
            let errorJson = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
            let msg = errorJson?["msg"] as? String
                ?? errorJson?["error_description"] as? String
                ?? "Apple sign-in failed"
            throw AuthError.serverError(msg)
        }
    }

    // MARK: - Google Sign-In (via ASWebAuthenticationSession)

    @MainActor
    func signInWithGoogle(anchor: ASPresentationAnchor) async throws {
        let redirectURI = "com.memorezar.app://callback"
        let authURL = URL(string: "\(SupabaseConfig.projectURL)/auth/v1/authorize?provider=google&redirect_to=\(redirectURI)")!

        let callbackURL: URL = try await withCheckedThrowingContinuation { continuation in
            let session = ASWebAuthenticationSession(
                url: authURL,
                callbackURLScheme: "com.memorezar.app"
            ) { url, error in
                if let error = error {
                    if (error as NSError).code == ASWebAuthenticationSessionError.canceledLogin.rawValue {
                        continuation.resume(throwing: AuthError.cancelled)
                    } else {
                        continuation.resume(throwing: AuthError.networkError)
                    }
                    return
                }
                guard let url = url else {
                    continuation.resume(throwing: AuthError.invalidResponse)
                    return
                }
                continuation.resume(returning: url)
            }
            session.presentationContextProvider = self
            session.prefersEphemeralWebBrowserSession = false
            session.start()
        }

        // Parse tokens from callback URL fragment (Supabase returns them in the hash)
        guard let fragment = callbackURL.fragment else {
            throw AuthError.invalidResponse
        }

        var params: [String: String] = [:]
        for pair in fragment.split(separator: "&") {
            let parts = pair.split(separator: "=", maxSplits: 1)
            if parts.count == 2 {
                params[String(parts[0])] = String(parts[1]).removingPercentEncoding ?? String(parts[1])
            }
        }

        guard let accessToken = params["access_token"],
              let refreshToken = params["refresh_token"] else {
            throw AuthError.invalidResponse
        }

        // Store tokens
        KeychainHelper.save(key: "access_token", string: accessToken)
        KeychainHelper.save(key: "refresh_token", string: refreshToken)

        // Fetch user info from the token
        let user = try await fetchUser(accessToken: accessToken)
        await MainActor.run {
            self.currentUser = user
            self.startRefreshTimer()
        }
        if let userData = try? JSONEncoder().encode(user) {
            KeychainHelper.save(key: "user_data", data: userData)
        }
    }

    // MARK: - Sign Out

    func signOut() {
        KeychainHelper.delete(key: "access_token")
        KeychainHelper.delete(key: "refresh_token")
        KeychainHelper.delete(key: "user_data")
        refreshTimer?.invalidate()
        refreshTimer = nil
        currentUser = nil
        errorMessage = nil
    }

    // MARK: - Token Refresh

    func refreshTokenIfNeeded() async {
        guard let refreshToken = refreshToken else { return }

        let url = URL(string: "\(SupabaseConfig.projectURL)/auth/v1/token?grant_type=refresh_token")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue(SupabaseConfig.anonKey, forHTTPHeaderField: "apikey")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body = ["refresh_token": refreshToken]
        request.httpBody = try? JSONEncoder().encode(body)

        do {
            let (data, response) = try await session.data(for: request)
            guard let httpResponse = response as? HTTPURLResponse,
                  (200...299).contains(httpResponse.statusCode),
                  let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                // Refresh failed — sign out
                await MainActor.run { self.signOut() }
                return
            }
            try handleAuthResponse(json: json)
        } catch {
            print("[AuthService] Token refresh failed: \(error)")
        }
    }

    // MARK: - Helpers

    private func handleAuthResponse(json: [String: Any], displayName: String? = nil) throws {
        guard let accessToken = json["access_token"] as? String,
              let refreshToken = json["refresh_token"] as? String else {
            throw AuthError.invalidResponse
        }

        KeychainHelper.save(key: "access_token", string: accessToken)
        KeychainHelper.save(key: "refresh_token", string: refreshToken)

        // Extract user from response
        let userJson = json["user"] as? [String: Any]
        let userId = userJson?["id"] as? String ?? ""
        let email = userJson?["email"] as? String
        let userMeta = userJson?["user_metadata"] as? [String: Any]
        let name = displayName
            ?? userMeta?["display_name"] as? String
            ?? userMeta?["full_name"] as? String
            ?? userMeta?["name"] as? String

        let user = AuthUser(id: userId, email: email, displayName: name)

        if let userData = try? JSONEncoder().encode(user) {
            KeychainHelper.save(key: "user_data", data: userData)
        }

        DispatchQueue.main.async {
            self.currentUser = user
            self.startRefreshTimer()
        }
    }

    private func fetchUser(accessToken: String) async throws -> AuthUser {
        let url = URL(string: "\(SupabaseConfig.projectURL)/auth/v1/user")!
        var request = URLRequest(url: url)
        request.setValue(SupabaseConfig.anonKey, forHTTPHeaderField: "apikey")
        request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode),
              let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw AuthError.invalidResponse
        }

        let userId = json["id"] as? String ?? ""
        let email = json["email"] as? String
        let userMeta = json["user_metadata"] as? [String: Any]
        let name = userMeta?["display_name"] as? String
            ?? userMeta?["full_name"] as? String
            ?? userMeta?["name"] as? String

        return AuthUser(id: userId, email: email, displayName: name)
    }

    private func startRefreshTimer() {
        refreshTimer?.invalidate()
        // Refresh every 50 minutes (Supabase tokens expire in 1 hour)
        refreshTimer = Timer.scheduledTimer(withTimeInterval: 50 * 60, repeats: true) { [weak self] _ in
            Task { await self?.refreshTokenIfNeeded() }
        }
    }
}

// MARK: - ASWebAuthenticationPresentationContextProviding

extension AuthService: ASWebAuthenticationPresentationContextProviding {
    func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        guard let scene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
              let window = scene.windows.first else {
            return ASPresentationAnchor()
        }
        return window
    }
}

// MARK: - Errors

enum AuthError: LocalizedError {
    case networkError
    case invalidResponse
    case serverError(String)
    case confirmationRequired
    case cancelled

    var errorDescription: String? {
        switch self {
        case .networkError: return "Network error. Check your connection."
        case .invalidResponse: return "Unexpected server response."
        case .serverError(let msg): return msg
        case .confirmationRequired: return "Check your email to confirm your account."
        case .cancelled: return nil
        }
    }
}

File: memorezar-android/app/src/main/java/com/memorezar/app/data/services/AuthService.kt
package com.memorezar.app.data.services

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import com.memorezar.app.data.models.AuthUser
import dagger.hilt.android.qualifiers.ApplicationContext
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.request.header
import io.ktor.client.request.post
import io.ktor.client.request.get
import io.ktor.client.request.setBody
import io.ktor.client.statement.HttpResponse
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.http.isSuccess
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import javax.inject.Inject
import javax.inject.Singleton

private const val TAG = "AuthService"
private const val REFRESH_INTERVAL_MS = 50 * 60 * 1000L // 50 minutes

@Singleton
class AuthService @Inject constructor(
    @ApplicationContext private val appContext: Context,
    private val httpClient: HttpClient
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private val json = Json { ignoreUnknownKeys = true }

    private val _currentUser = MutableStateFlow<AuthUser?>(null)
    val currentUser: StateFlow<AuthUser?> = _currentUser.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    val isSignedIn: Boolean get() = _currentUser.value != null

    val accessToken: String? get() = securePrefs.getString("access_token", null)
    private val refreshToken: String? get() = securePrefs.getString("refresh_token", null)

    private var refreshJob: Job? = null

    // Encrypted storage (replaces iOS Keychain)
    private val securePrefs: SharedPreferences by lazy {
        val masterKey = MasterKey.Builder(appContext)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
        try {
            EncryptedSharedPreferences.create(
                appContext,
                "memorezar_secure_prefs",
                masterKey,
                EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
            )
        } catch (_: Exception) {
            // Corrupted prefs (e.g. after uninstall/reinstall with stale KeyStore key)
            appContext.getSharedPreferences("memorezar_secure_prefs", Context.MODE_PRIVATE).edit().clear().apply()
            val prefsFile = java.io.File(appContext.filesDir.parent, "shared_prefs/memorezar_secure_prefs.xml")
            prefsFile.delete()
            EncryptedSharedPreferences.create(
                appContext,
                "memorezar_secure_prefs",
                masterKey,
                EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
            )
        }
    }

    // MARK: - Session Restore

    fun restoreSession() {
        val token = accessToken ?: return
        val userData = securePrefs.getString("user_data", null)
        if (userData != null) {
            try {
                _currentUser.value = json.decodeFromString<AuthUser>(userData)
                startRefreshTimer()
            } catch (e: Exception) {
                Log.w(TAG, "Failed to decode stored user: ${e.message}")
            }
        }
        // Silently refresh
        scope.launch { refreshTokenIfNeeded() }
    }

    // MARK: - Email/Password

    suspend fun signUpWithEmail(email: String, password: String, name: String) {
        _isLoading.value = true
        _errorMessage.value = null
        try {
            val response: HttpResponse = httpClient.post("${SupabaseConfig.PROJECT_URL}/auth/v1/signup") {
                header("apikey", SupabaseConfig.ANON_KEY)
                contentType(ContentType.Application.Json)
                setBody("""{"email":"$email","password":"$password","data":{"display_name":"$name"}}""")
            }
            if (response.status.isSuccess()) {
                val body: String = response.body()
                val jsonObj = json.parseToJsonElement(body).jsonObject
                if (jsonObj.containsKey("access_token")) {
                    handleAuthResponse(jsonObj, displayName = name)
                } else {
                    throw AuthError.ConfirmationRequired
                }
            } else {
                val errorMsg = parseErrorMessage(response) ?: "Sign up failed"
                throw AuthError.ServerError(errorMsg)
            }
        } catch (e: AuthError) {
            _errorMessage.value = e.message
            throw e
        } catch (e: Exception) {
            _errorMessage.value = "Network error. Check your connection."
            throw AuthError.NetworkError
        } finally {
            _isLoading.value = false
        }
    }

    suspend fun signInWithEmail(email: String, password: String) {
        _isLoading.value = true
        _errorMessage.value = null
        try {
            val response: HttpResponse = httpClient.post(
                "${SupabaseConfig.PROJECT_URL}/auth/v1/token?grant_type=password"
            ) {
                header("apikey", SupabaseConfig.ANON_KEY)
                contentType(ContentType.Application.Json)
                setBody("""{"email":"$email","password":"$password"}""")
            }
            if (response.status.isSuccess()) {
                val body: String = response.body()
                handleAuthResponse(json.parseToJsonElement(body).jsonObject)
            } else {
                val errorMsg = parseErrorMessage(response) ?: "Invalid email or password"
                throw AuthError.ServerError(errorMsg)
            }
        } catch (e: AuthError) {
            _errorMessage.value = e.message
            throw e
        } catch (e: Exception) {
            _errorMessage.value = "Network error. Check your connection."
            throw AuthError.NetworkError
        } finally {
            _isLoading.value = false
        }
    }

    // MARK: - Apple Sign-In (via Supabase id_token)

    suspend fun signInWithApple(idToken: String, nonce: String?) {
        _isLoading.value = true
        _errorMessage.value = null
        try {
            val bodyStr = buildString {
                append("""{"provider":"apple","id_token":"$idToken"""")
                if (nonce != null) append(""","nonce":"$nonce"""")
                append("}")
            }
            val response: HttpResponse = httpClient.post(
                "${SupabaseConfig.PROJECT_URL}/auth/v1/token?grant_type=id_token"
            ) {
                header("apikey", SupabaseConfig.ANON_KEY)
                contentType(ContentType.Application.Json)
                setBody(bodyStr)
            }
            if (response.status.isSuccess()) {
                val body: String = response.body()
                handleAuthResponse(json.parseToJsonElement(body).jsonObject)
            } else {
                val errorMsg = parseErrorMessage(response) ?: "Apple sign-in failed"
                throw AuthError.ServerError(errorMsg)
            }
        } catch (e: AuthError) {
            _errorMessage.value = e.message
            throw e
        } catch (e: Exception) {
            _errorMessage.value = "Network error. Check your connection."
            throw AuthError.NetworkError
        } finally {
            _isLoading.value = false
        }
    }

    // MARK: - Google Sign-In (via OAuth redirect)
    // On Android, Google Sign-In uses Credential Manager.
    // The callback URL fragment parsing is the same.

    suspend fun handleOAuthCallback(fragment: String) {
        val params = mutableMapOf<String, String>()
        for (pair in fragment.split("&")) {
            val parts = pair.split("=", limit = 2)
            if (parts.size == 2) {
                params[parts[0]] = java.net.URLDecoder.decode(parts[1], "UTF-8")
            }
        }

        val token = params["access_token"] ?: throw AuthError.InvalidResponse
        val refresh = params["refresh_token"] ?: throw AuthError.InvalidResponse

        securePrefs.edit()
            .putString("access_token", token)
            .putString("refresh_token", refresh)
            .apply()

        val user = fetchUser(token)
        _currentUser.value = user
        saveUserData(user)
        startRefreshTimer()
    }

    fun getGoogleOAuthURL(): String {
        val redirectURI = "com.memorezar.app://callback"
        return "${SupabaseConfig.PROJECT_URL}/auth/v1/authorize?provider=google&redirect_to=$redirectURI"
    }

    // MARK: - Sign Out

    fun signOut() {
        securePrefs.edit()
            .remove("access_token")
            .remove("refresh_token")
            .remove("user_data")
            .apply()
        refreshJob?.cancel()
        refreshJob = null
        _currentUser.value = null
        _errorMessage.value = null
    }

    // MARK: - Token Refresh

    suspend fun refreshTokenIfNeeded() {
        val token = refreshToken ?: return
        try {
            val response: HttpResponse = httpClient.post(
                "${SupabaseConfig.PROJECT_URL}/auth/v1/token?grant_type=refresh_token"
            ) {
                header("apikey", SupabaseConfig.ANON_KEY)
                contentType(ContentType.Application.Json)
                setBody("""{"refresh_token":"$token"}""")
            }
            if (response.status.isSuccess()) {
                val body: String = response.body()
                handleAuthResponse(json.parseToJsonElement(body).jsonObject)
            } else {
                // Refresh failed — sign out
                signOut()
            }
        } catch (e: Exception) {
            Log.w(TAG, "Token refresh failed: ${e.message}")
        }
    }

    // MARK: - Account Deletion

    /**
     * Calls the `delete-user-account` Edge Function with the current user's
     * JWT. Returns null on success, an error string on failure. The function
     * cascade-deletes recordings (rows + audio files), user_backups,
     * support_tickets, and the auth.users row.
     */
    suspend fun deleteAccountOnServer(): String? {
        val token = accessToken ?: return "Not signed in"
        return try {
            val response: HttpResponse = httpClient.post(
                "${SupabaseConfig.PROJECT_URL}/functions/v1/delete-user-account"
            ) {
                header("apikey", SupabaseConfig.ANON_KEY)
                header("Authorization", "Bearer $token")
                contentType(ContentType.Application.Json)
                setBody("{}")
            }
            if (response.status.isSuccess()) {
                null
            } else {
                val body: String = try { response.body() } catch (_: Exception) { "" }
                "HTTP ${response.status.value}: $body"
            }
        } catch (e: Exception) {
            e.message ?: "Unknown error"
        }
    }

    // MARK: - Helpers

    /** Build standard Supabase headers with current auth token. */
    fun headers(): Map<String, String> {
        val token = accessToken ?: SupabaseConfig.ANON_KEY
        return mapOf(
            "apikey" to SupabaseConfig.ANON_KEY,
            "Authorization" to "Bearer $token",
            "Content-Type" to "application/json"
        )
    }

    private fun handleAuthResponse(jsonObj: JsonObject, displayName: String? = null) {
        val tokenStr = jsonObj["access_token"]?.jsonPrimitive?.content ?: throw AuthError.InvalidResponse
        val refreshStr = jsonObj["refresh_token"]?.jsonPrimitive?.content ?: throw AuthError.InvalidResponse

        securePrefs.edit()
            .putString("access_token", tokenStr)
            .putString("refresh_token", refreshStr)
            .apply()

        val userObj = jsonObj["user"]?.jsonObject
        val userId = userObj?.get("id")?.jsonPrimitive?.content ?: ""
        val email = userObj?.get("email")?.jsonPrimitive?.content
        val meta = userObj?.get("user_metadata")?.jsonObject
        val name = displayName
            ?: meta?.get("display_name")?.jsonPrimitive?.content
            ?: meta?.get("full_name")?.jsonPrimitive?.content
            ?: meta?.get("name")?.jsonPrimitive?.content

        val user = AuthUser(id = userId, email = email, displayName = name)
        saveUserData(user)
        _currentUser.value = user
        startRefreshTimer()
    }

    private suspend fun fetchUser(accessToken: String): AuthUser {
        val response: HttpResponse = httpClient.get("${SupabaseConfig.PROJECT_URL}/auth/v1/user") {
            header("apikey", SupabaseConfig.ANON_KEY)
            header("Authorization", "Bearer $accessToken")
        }
        if (!response.status.isSuccess()) throw AuthError.InvalidResponse

        val body: String = response.body()
        val jsonObj = json.parseToJsonElement(body).jsonObject
        val userId = jsonObj["id"]?.jsonPrimitive?.content ?: ""
        val email = jsonObj["email"]?.jsonPrimitive?.content
        val meta = jsonObj["user_metadata"]?.jsonObject
        val name = meta?.get("display_name")?.jsonPrimitive?.content
            ?: meta?.get("full_name")?.jsonPrimitive?.content
            ?: meta?.get("name")?.jsonPrimitive?.content

        return AuthUser(id = userId, email = email, displayName = name)
    }

    private fun saveUserData(user: AuthUser) {
        val encoded = json.encodeToString(AuthUser.serializer(), user)
        securePrefs.edit().putString("user_data", encoded).apply()
    }

    private fun startRefreshTimer() {
        refreshJob?.cancel()
        refreshJob = scope.launch {
            while (true) {
                delay(REFRESH_INTERVAL_MS)
                refreshTokenIfNeeded()
            }
        }
    }

    private suspend fun parseErrorMessage(response: HttpResponse): String? {
        return try {
            val body: String = response.body()
            val obj = json.parseToJsonElement(body).jsonObject
            obj["msg"]?.jsonPrimitive?.content
                ?: obj["error_description"]?.jsonPrimitive?.content
        } catch (_: Exception) {
            null
        }
    }
}

// MARK: - Errors

sealed class AuthError(override val message: String) : Exception(message) {
    object NetworkError : AuthError("Network error. Check your connection.")
    object InvalidResponse : AuthError("Unexpected server response.")
    data class ServerError(val msg: String) : AuthError(msg)
    object ConfirmationRequired : AuthError("Check your email to confirm your account.")
    object Cancelled : AuthError("Sign-in was cancelled.")
}
